lxd/request/request.go

```diff
@@ -78,7 +78,8 @@ func CountStartedRequest(r *http.Request) {
 // The callback function is used to mark the request as completed for the API metrics.
 func MetricsCallback(request *http.Request, result metrics.RequestResult) {
 	callback, err := GetCtxValue[func(metrics.RequestResult)](request.Context(), MetricsCallbackFunc)
-	if err != nil && (strings.HasPrefix(request.URL.Path, "/1.0") || request.URL.Path == "/") {
+	// If host is a custom socket the request is coming from the /dev/lxd socket.
+	if err != nil && strings.HasPrefix(request.URL.Path, "/1.0") && request.Host != "custom.socket" {
```

Comment:
AFAIK, this `custom.socket` host can be basically anything:

```
$ lxc exec c1 -- curl -s --unix-socket /dev/lxd/sock http://custom.socket/1.0
{"state":"Started","api_version":"1.0","instance_type":"container","location":"sdeziel-lemur"}
$ lxc exec c1 -- curl -s --unix-socket /dev/lxd/sock http://foo.bar/1.0
{"state":"Started","api_version":"1.0","instance_type":"container","location":"sdeziel-lemur"}
```

I think @markylaing suggested checking for `@devlxd` somewhere in the request struct(?).

Reply:
Yes, but it unfortunately did not work :(

The `RemoteAddr` value followed the pattern of `vm(104676174):2672481575` instead of being `@devlxd`.

I was afraid checking for `custom.socket` could be too broad, but I think the alternative would be using a regex to check if `RemoteAddr` follows the expected pattern fot vm remotes. Or remove this log message entirely.

Reply:
`vm(104676174):2672481575` looks like a `vsock` address so maybe that is a good way to distinguish them for the VM case? The `@devlxd` request address might be just for containers?

Reply:
@hamistao do you know why this function is being called at all for devlxd?

Reply:
Because it uses some of the same `Response` types as the regular API. See [this](https://github.com/hamistao/lxd/blob/improve_reboot_experience/lxd/response/response.go#L94) line.

Reply:
> The @devlxd request address might be just for containers

If this is the case, should we worry the existing checks for `@devlxd` may be flawed?

> vm(104676174):2672481575 looks like a vsock address so maybe that is a good way to distinguish them for the VM case

Cool, that works for me as well :D

Reply:
It seems to be like in a single API handler for devlxd we override the `RemoteAddr` with `devlxdRemoteAddress` and for all the others we do not do that.

Then there are checks for `RemoteAddr` being `devlxdRemoteAddress` to prevent access to things or to classify the request and thus this isn't working correctly for all devlxd endpoints.

I also don't really like us overwriting the `RemoteAddr` with `devlxdRemoteAddress` because we lose the original remote identifier.

Instead I think instead of this we should remove `devlxdRemoteAddress` and instead have the devlxd API server set something into the request context to indicate its a devlxd request.

We already have `request.CtxProtocol` which is used to indicate the authentication method, perhaps this would be an appropriate place to have "devlxd" as another authentication method. 

Then we can update the places that check for `devlxdRemoteAddress`  to check for `request.CtxProtocol` for as "devlxd".

What do you think @markylaing ?

Reply:
@tomponline
If @markylaing agrees, I can implement your suggestion as a part of PR since it will help with the problem at hand as well.

Reply:
> Instead I think instead of this we should remove devlxdRemoteAddress and instead have the devlxd API server set something into the request context to indicate its a devlxd request.

Yep this sounds better than setting the `RemoteAddr`. @hamistao you can add the new authentication method under `lxd/auth/types.go`.

Reply:
That being said. If we don't want these to count towards metrics at all... would it be better to just delete the metrics callback in the devlxd request hoist function?

Reply:
> you can add the new authentication method under lxd/auth/types.go

So this authentication method would be just checking the request context for whether the request is a devlxd request?

> Or better: don't set it when setting up the devlxd handlers.

It is not being set up, this log message is triggered by there being no callback function in the request context.

Reply:
> It is not being set up, this log message is triggered by there being no callback function in the request context.

Ah that makes sense, sorry my misunderstanding.

> So this authentication method would be just checking the request context for whether the request is a devlxd request?

Sorry this is overloading the term "method". The authentication methods are string constants: https://github.com/canonical/lxd/blob/dbefa5e712d5a026893957e79b277485b770eb20/lxd/auth/types.go#L11-L29 

more are defined in the `api` package as well: https://github.com/canonical/lxd/blob/dbefa5e712d5a026893957e79b277485b770eb20/shared/api/auth.go#L3-L9

but we'll keep the `devlxd` authentication method under `lxd/auth` because it's internal to LXD.

Currently, setting `r.RemoteAddr` is only being performed in the devlxd image export handler to distinguish between devlxd and main API queries in the main image export handler. We should:
1. Remove `devlxdRemoteAddress`.
2. Remove setting `r.RemoteAddr` in the devlxd image export handler.
3. Add a new authentication method constant for devlxd in `lxd/auth/types.go` with key `request.CtxProtocol`.
4. Set the authentication method in the request context in devlxd's `hoistReq` function.
5. Where we previously checked `r.RemoteAddr`, now call `(lxd/auth).GetAuthenticationMethodFromCtx` to get the authentication method.
6. Skip logging request not counted based on the value returned by `(lxd/auth).GetAuthenticationMethodFromCtx`.